MahoCommerce/maho
# Added ability to view orders paid with removed payment method without breaking

# Description

Maho currently can't handle deleted/removed payment methods in orders in the frontend as well as in the adminhtml.

Example: An external payment module for the payment method `Amazon Pay` had been installed. Some customers used this payment method for their orders in the past. The module was then removed again because a different payment method was to be introduced instead.

A customer who used such removed payment method currrently will get this error when viewing the order in the customer account:
```
There has been an error processing your request

The requested Payment Method is not available.

Trace:
#0 /var/www/html/app/code/core/Mage/Payment/Model/Info.php(101): Mage::throwException('The requested P...')
#1 /var/www/html/app/code/core/Mage/Payment/Helper/Data.php(131): Mage_Payment_Model_Info->getMethodInstance()
#2 /var/www/html/app/code/core/Mage/Sales/Block/Order/Info.php(47): Mage_Payment_Helper_Data->getInfoBlock(Object(Mage_Sales_Model_Order_Payment))
#3 /var/www/html/app/code/core/Mage/Core/Block/Abstract.php(299): Mage_Sales_Block_Order_Info->_prepareLayout()
#4 /var/www/html/app/code/core/Mage/Core/Model/Layout.php(474): Mage_Core_Block_Abstract->setLayout(Object(Mage_Core_Model_Layout))
#5 /var/www/html/app/code/core/Mage/Core/Model/Layout.php(490): Mage_Core_Model_Layout->createBlock('sales/order_inf...', 'sales.order.inf...')
#6 /var/www/html/app/code/core/Mage/Core/Model/Layout.php(234): Mage_Core_Model_Layout->addBlock('sales/order_inf...', 'sales.order.inf...')
#7 /var/www/html/app/code/core/Mage/Core/Model/Layout.php(200): Mage_Core_Model_Layout->_generateBlock(Object(Mage_Core_Model_Layout_Element), Object(Mage_Core_Model_Layout_Element))
#8 /var/www/html/app/code/core/Mage/Core/Model/Layout.php(205): Mage_Core_Model_Layout->generateBlocks(Object(Mage_Core_Model_Layout_Element))
#9 /var/www/html/app/code/core/Mage/Core/Controller/Varien/Action.php(348): Mage_Core_Model_Layout->generateBlocks()
#10 /var/www/html/app/code/core/Mage/Core/Controller/Varien/Action.php(259): Mage_Core_Controller_Varien_Action->generateLayoutBlocks()
#11 /var/www/html/app/code/core/Mage/Sales/Controller/Abstract.php(51): Mage_Core_Controller_Varien_Action->loadLayout()
#12 /var/www/html/app/code/core/Mage/Sales/Controller/Abstract.php(93): Mage_Sales_Controller_Abstract->_viewAction()
#13 /var/www/html/app/code/core/Mage/Core/Controller/Varien/Action.php(424): Mage_Sales_Controller_Abstract->viewAction()
#14 /var/www/html/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(255): Mage_Core_Controller_Varien_Action->dispatch('view')
#15 /var/www/html/app/code/core/Mage/Core/Controller/Varien/Front.php(181): Mage_Core_Controller_Varien_Router_Standard->match(Object(Mage_Core_Controller_Request_Http))
#16 /var/www/html/app/code/core/Mage/Core/Model/App.php(358): Mage_Core_Controller_Varien_Front->dispatch()
#17 /var/www/html/app/Mage.php(700): Mage_Core_Model_App->run(Array)
#18 /var/www/html/pub/index.php(59): Mage::run('', 'store')
#19 {main}
```

This PR handles removed payment methods in the frontend e.g. on customer account order view pages as well as in the adminhtml e.g. on sales order view pages so they will successfully render again.

# Type of change

Fix

# How can this fix be tested

1. Checkout the main branch (without this fix).
2. Create a customer account.
3. Place an order.
4. View the order in the customer account.
5. The order view page will render successfully and displays, among other things, the payment method selected when ordering correctly.
6. Change the payment method of the order in `sales_flat_order_payment` database table (column `method`) to a non existent payment method (to simulate a removed payment method), e.g. `some_obsolete_payment_method`.
7. Reload the customer account order view page and see the above error message.
8. Checkout the branch of this PR including the fix.
9. Reload the customer account order view page.
10. The page will render successfully again and displays, among other things, at least the payment method code (e.g. `some_obsolete_payment_method`) in the payment information block.
11. A notice is logged so that this case can be tracked if necessary.
12. View the oder in adminhtml.
13. The sales oder view page will render successfully again and displays, among other things, at least the payment method code (e.g. `some_obsolete_payment_method`) in the payment information block.
